mkosi/archive.py

```diff
@@ -63,6 +67,72 @@ def make_tar(src: Path, dst: Path, *, tools: Path = Path("/"), sandbox: Sequence
         )
 
 
+def make_oci(context: Context, dst: Path) -> None:
+    ca_store = dst / "blobs" / "sha256"
+    ca_store.mkdir(parents=True)
+
+    layer_archive = context.staging / context.config.output_tar
+    layer_digest = hash_file(layer_archive)
+    # Should we create a symlink at the old path such that the tar can still be used?
+    layer_archive.rename(ca_store / layer_digest)
+
+    oci_config = {
+        # TODO Add more metadata? Created timestamp, history, config...
```

Comment:
timestamp makes sense, though that should probably use `SOURCE_DATE_EPOCH`. Are there any fields that could use the image version or image id? 

No idea what history and config are.

Reply:
> timestamp makes sense, though that should probably use `SOURCE_DATE_EPOCH`.

Good idea.

> Are there any fields that could use the image version or image id?

There is `os.version` though that is for the host system on which this should run (which is also the case for `os`).

Otherwise we have labels/annotations, [many of which are standardized](https://github.com/opencontainers/image-spec/blob/365fa41472a19107b0b4cbc17e141a9b97f9b66e/annotations.md#pre-defined-annotation-keys). For us `org.opencontainers.image.version` would work. There is also `org.opencontainers.image.base.name` but that is actually for base images in the Docker sense, not our base distribution.

> No idea what history and config are.

Config can be used to preset information used when running the image. E.g. an entrypoint/cmd, working dir, environment variables. One idea would be to set `/sbin/init` here though I am not sure if that is the best idea. If we do not set anything at least the entrypoint/cmd always has to be given explicitly. Working dir seems to default to `/` and env is empty apart from what the container runtime sets.
(Users are still able to create a new image building upon the oci output from mkosi and set these afterwards. Though I am not sure about the UX for docker - might be that they have to explicitly load the oci image into the local registry first. With podman this is as simple as `FROM oci:mkosi.output/image\nCMD /usr/bin/bash`)

History is used to log the stuff applied to the image, e.g. commands run, files added etc. Dockerhub seem to show them when viewing details about specific image version, e.g. https://hub.docker.com/layers/library/fedora/40/images/sha256-e7c8351c0292634436207cd10654eff5361153f0617e71c8fbd6e8dcf8d6f826?context=explore. These mostly seem to use the syntax from Dockerfiles though the spec does not set any restrictions on them. I am not sure what would happen if one where to just write random stuff in there.
We could save all invoked commands and store them in the history but that seems rather cumbersome and not worthwhile.

Reply:
> Though I am not sure about the UX for docker - might be that they have to explicitly load the oci image into the local registry first. With podman this is as simple as `FROM oci:mkosi.output/image\nCMD /usr/bin/bash`)

I just checked how this works with docker. Unless I missed something one has to run `tar -c . | docker import - my-mkosi-oci` before being able to use the image. Docker (or buildx) does not support simply using `oci:...` as an image specifier but instead interprets it as an image with name "oci".

Reply:
History and config seem a bit pointless or too far out, but setting `org.opencontainers.image.base.name` might make sense. Wouldn't the images produced by mkosi be for all intents and purpses base images?

Reply:
Yes though the base name annotation is for the other way around. It should refer to the image upon which this image is based. E.g. if I create an image based on `registry.fedoraproject.org/fedora:latest` then that would saved in the annotations.

Reply:
I tried out setting a `Cmd` in config such that `/sbin/init` is started automatically, I think that might be a reasonable thing to do but we can also drop it. I am a bit toying around with what would make sense.

I also added a single history field for testing with a `comment` and `created_by` field. The comment does not seem to show up but the created_by field is at least shown in Docker Hub for my demo image: https://hub.docker.com/layers/septatrix/mkosi-oci-fedora-test/latest/images/sha256-a81676033c96a8de79a3441640fbfb3ca14da09276e1409281e7ce02ecc44e55?context=explore